Adding FRD file header (DAQ)

smorovic opened the pull request:
#### PR description:

Addition of file header to Fed Raw Data format. This header contains metadata description of the content (file size without header and number of events, LS and versioning). Aim is to replace separate JSON file which can take a lot of space in ramdisk with 2 MB huge pages (minimum per file).

File broker will switch protocol to v2 when detecting raw files with header, and Input source assumes i in this case . In addition there is also format autodetection when using fileListMode and in FRDSource, as well as generation support for the new format in the fake BU module.

Additional optimization is added: input source will, in case of using a single reader thread with double buffering (default mode), open raw file only once and reuse the same FD to read file sequentially until done (reducing number file operations performed).

#### PR validation:

This has been developed and tested in DAQ3VAL system.

smorovic committed:
New Fed Raw Data (FRD) format with a fixed size header. It contains
identification and version fields, as well as metadata (file size and
number of events), which are used for consistency checks in HLT nodes.

In case of HLT (NFS) mode this is activated only when using file broker
protocol v2. In this case accompanying JSON file is no longer present. In
case of fileListMode of the FedRawDataSource and with FRDStramSource,
it is autodetected based on reading the raw file.

Commit includes a bugfix: use separate variable for reading second field
in JSON file (affecting protocol with separate JSON index file; fortunately
the bug has not been hit in the production system.

Note: code-format-all has been executed and manually fixed due to changes
being made prior to reformatting campaign

smorovic committed:
uint32 is 4-byte aligned, uint64 is 8-byte aligned. Therefore added padding after uint16 as reserved 16-bit field.
Size is in either case 24, not 22. Use uint8 for consistency

smorovic committed:
abandon uint* and switch to uint*_t which is C++ standard. Other users
in Streamer should also be migrated eventually.

smorovic committed:
- fakeBU implementation of writing file header
- format update and change in initialization of the header structure

smorovic committed:
added lumisection to FRDF header

smorovic committed:
fixes to fake BU

smorovic committed:
fixed parsing of version from FRD header string-like representation

smorovic committed:
updating test script. FRDStreamSource now requires parameter which
specifies [ls,run] tuple

smorovic committed:
reuse file descriptor to read file. take care of adjusting buffer position when inheriting.
In case of 2 buffers (1 thread) only a single fd will be used with only
sequential reading

smorovic committed:
fix parsing of version numbers

smorovic committed:
update description

smorovic committed:
- use correct JSON format
- fix amount read into chunk in case of skipping header payload

smorovic committed:
pointer position needs to be increased in uint8 x size, not uint32 x size

smorovic committed:
include raw path in index json file

smorovic committed:
move FRDFileHeader definition into a separate header

smorovic committed:
revert change to FRDEventMEssage.cc

smorovic committed:
code-format-all

smorovic committed:
fix string size

cmsbuild commented:
@smorovic, CMSSW_11_0_X branch is closed for direct updates. cms-bot is going to move this PR to master branch.
In future, please use cmssw master branch to submit your changes.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27750/11395

- This PR adds an extra 212KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27750/11395/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27750/11395/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27750/11395/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27750/11395/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

smorovic committed:
applying code-format and code-checks patch

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27750/11396

- This PR adds an extra 124KB to repository

cmsbuild commented:
A new Pull Request was created by  @smorovic (Srecko Morovic) for master.

It involves the following packages:

EventFilter/Utilities
IOPool/Streamer


@perrotta, @smuzaffar, @Dr15Jones, @emeschi, @cmsbuild, @slava77, @mommsen can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @makortel, @wddgit this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 9b6c044233e359c8f3a1276168f4cbadd9900b7b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0dcc38/1970/summary.html

cmsbuild commented:
Comparison job queued.

Dr15Jones commented on EventFilter/Utilities/src/FedRawDataInputSource.cc:
```diff
@@ -992,8 +1051,15 @@ void FedRawDataInputSource::readSupervisor() {
 
-        InputFile* newInputFile = new InputFile(
-            evf::EvFDaqDirector::FileStatus::newFile, ls, rawFile, fileSize, neededChunks, eventsInNewFile, this);
+        InputFile* newInputFile = new InputFile(evf::EvFDaqDirector::FileStatus::newFile,
```
Even better, change `fileQueue_` so it holds `std::unique_ptr` (assuming the objects eventually get deleted).

smorovic replied:
Now InputFile objects are contained in std::unique_ptr.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0dcc38/1970/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2939508
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2939165
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 142 log files, 14 edm output root files, 34 DQM output files

smorovic commented on EventFilter/Utilities/src/FedRawDataInputSource.cc:
```diff
@@ -398,6 +398,8 @@ inline evf::EvFDaqDirector::FileStatus FedRawDataInputSource::getNextEvent() {
       }
     //immediately delete empty file
+    if (currentFile_->rawFd_ != -1)
+      close(currentFile_->rawFd_);
```
I will change it. A small caveat is, with current code it would mean deleting file before closing it, so that has to be rearranged. File deletion is in some cases done later (to make sure that file is present until after events are processed) so it's not done in InputFile class.

smorovic replied:
After another look, the object is always removed when the file is, so it can be done in destructor.

smorovic commented on EventFilter/Utilities/src/EvFDaqDirector.cc:
```diff
@@ -941,6 +943,187 @@ namespace evf {
+      }
+      edm::LogError("EvFDaqDirector") << "grabNextJsonFromRaw - failed to open output file -: " << jsonDestPath << " : "
+                                      << strerror(errno);
+      usleep(100000);
```
This is leftover from attempts to solve problems with NFS. I think we didn't see those errors after lookup cache NFS parameters were configured appropriately.

We might not need it, so I'll remove the sleep but leave retry logic (for now).

smorovic commented on EventFilter/Utilities/plugins/FRDStreamSource.cc:
```diff
@@ -37,6 +38,30 @@ bool FRDStreamSource::setRunAndEventInfo(edm::EventID& id,
   }
+  //look for FRD header at beginning of the file and skip it
+  if (fin_.tellg() == 0) {
+    const size_t buf_sz = sizeof(FRDFileHeader_v1);  //try to read v1 FRD header size
```
Good point. Will change it to use the struct directly on stack (assuming that should guarantee alignment).

smorovic committed:
fixes for comments by Dr15Jones part1:
- destructor for InputFile closing the file. Also realized a race
  condition with >=3 buffers when closing files, which is here resolved
- contexpr and std::array usage, removing need for preprocessor
  directives
- fixing potential alignment issue with FRDFileHeader_v1 class on stack
-

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27750/11430

- This PR adds an extra 144KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27750/11430/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27750/11430/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

smorovic committed:
- throwing exception on retry if file delete still fails
- additional fix for part1: delete file after closing

smorovic committed:
- unique_ptr used instead of InputFile pointer
- deletion of the file now in destructor of InputFile

smorovic committed:
bugfix: reading old format was broken when using smaller block size than
buffer(chunk) size with the new header handling

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27750/11445

- This PR adds an extra 148KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27750/11445/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27750/11445/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

smorovic committed:
applying code-format patch

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27750/11447

- This PR adds an extra 212KB to repository

cmsbuild commented:
Pull request #27750 was updated. @perrotta, @smuzaffar, @Dr15Jones, @emeschi, @cmsbuild, @slava77, @mommsen can you please check and sign again.

smorovic commented:
Thanks for comments. I retested due to amount of changes (and additional bugfix) and no issues were found.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 71cf55c33b63554c18d5a3f7915af52fcfea1972
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0dcc38/2016/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0dcc38/2016/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2939508
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2939166
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 142 log files, 14 edm output root files, 34 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/27750	71cf55c
- the modified code is not in the reco responsibility, related only to the data I/O in the online applications (even though it's in EventFilter/Utilities which holds some reco-related code as well)
- jenkins tests pass

emeschi commented:
+1

Dr15Jones commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1
